lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h

```diff
@@ -0,0 +1,377 @@
+/// each type of struct declared in the program.
+class StructFieldAccessManager {
+public:
+  /// enum used to represent different type of struct definitions
+  enum StructDefinitionType {
+    SDT_GlobalStruct,
+    SDT_GlobalStructPtr,
+    SDT_GlobalStructPtrPtr,
+    SDT_LocalStruct,
+    SDT_LocalStructPtr,
+    SDT_LocalStructPtrPtr
+  };
+
+  /// enum used to count the corner cases in the program for future
+  /// consideration
+  enum DebugStats {
+    DS_StructPtrPtr,
+    DS_FuncArgValue,
+    DS_FuncArgNotDefined,
+    DS_GepPassedIntoIndirectFunc,
+    DS_GepPassedIntoBitcast,
+    DS_GepUnknownUse,
+    DS_UserNotInstructionNorOperator,
+    DS_NoAccess,
+    DS_PassedIntoOutsideFunction,
+    DS_GepUsedOnStructPtr,
+    DS_UnknownUsesOnStructPtr,
+    DS_FilterColdStructs,
+    DS_MaxNumStats
+  };
+
+  StructFieldAccessManager(const Module &M,
+                           function_ref<BlockFrequencyInfo *(Function &)> LBFI)
+      : CurrentModule(M), LookupBFI(LBFI),
+        StatCounts(DebugStats::DS_MaxNumStats) {
+    HotnessAnalyzer = new StructHotnessAnalyzer;
+  };
+
+  ~StructFieldAccessManager() { delete HotnessAnalyzer; }
+
+  /// Check if the struct type is created before; if not, create a new
+  /// StructFieldAccessInfo object for it
+  StructFieldAccessInfo *
+  createOrGetStructFieldAccessInfo(const Type *T,
+                                   const StructDefinitionType ST);
+
+  /// Retrieve the pointer to the previous created StructFieldAccessInfo object
+  /// for the type
+  StructFieldAccessInfo *getStructFieldAccessInfo(const Type *T) const;
+
+  /// Retrieve execution count for a basic block
+  Optional<uint64_t> getExecutionCount(const BasicBlock *BB) const {
+    Function *Func = const_cast<Function *>(BB->getParent());
+    return LookupBFI(*Func)->getBlockProfileCount(BB);
+  }
+
+  /// Retrive a pair of information if the instruction is accessing any struct
+  /// type and field number
+  Optional<StructInfoMapPairType>
+  getFieldAccessOnInstruction(const Instruction *I) const;
+
+  // Summarizes all CallInst and InvokeInst into function declarations
+  void summarizeFunctionCalls();
+  // Apply some filters to reduce the number of struct in analysis
+  void applyFiltersToStructs();
+
+  /// Print all accesses of all struct types defined in the program
+  void debugPrintAllStructAccesses();
+
+  /// Print the IR of the module with annotated information about struct access
+  void debugPrintAnnotatedModule();
+
+  /// Increment stats for one category
+  void addStats(unsigned Category) { StatCounts[Category]++; }
+  /// Print a brief stats of struct access
+  void printStats();
+
+private:
+  const Module &CurrentModule;
+
+  /// Function reference that is used to retrive execution count for basic block
+  function_ref<BlockFrequencyInfo *(Function &)> LookupBFI;
+
+  /// A map storing access info of all structs
+  std::unordered_map<const StructType *, StructFieldAccessInfo *>
```

Comment:
See if you can use a DenseMap instead (http://llvm.org/docs/ProgrammersManual.html#llvm-adt-densemap-h).